bootstore/src/schemes/v0/peer.rs

```diff
@@ -1217,80 +1419,60 @@ mod tests {
 
         // Wipe the learner ledger, restart the learner and instruct it to
         // relearn its share, and ensure that the neither generation number gets
-        // bumped because persistence doesn't occur.
-        learner_handle.shutdown().await.unwrap();
-        learner_jh.await.unwrap();
-        std::fs::remove_file(&learner_config.fsm_state_ledger_paths[0])
-            .unwrap();
-        let (mut learner, learner_handle) =
-            Node::new(learner_config.clone(), &log).await;
-        let learner_jh = tokio::spawn(async move {
-            learner.run().await;
-        });
-        let _ = learner_handle.load_peer_addresses(addrs.clone()).await;
-        learner_handle.init_learner().await.unwrap();
+        // bumped because persistence doesn't occur. But for that to happen
+        // we need to make sure the learner asks the same peer. The choice of
+        // which peer is somewhat implementation dependent, so to get around
+        // that we stop the other peer until after restarting the learner.
+
+        // Infer the correct peer based on which generation number didn't get
+        // bumped and stop it.
+        let non_learner_peer = if peer0_gen_new == peer0_gen { 0 } else { 1 };
+        nodes.shutdown_node(non_learner_peer).await;
+
+        // Now we can stop the learner, wipe its ledger, and restart it.
+        nodes.shutdown_learner(true).await;
+        nodes.start_learner().await;
+
+        // Now restart the other peer and tell the learner to relearn its share
+        nodes.start_node(non_learner_peer).await;
```

Comment:
I believe this is non-deterministic, and we are just getting lucky. When `init_learner()` is called it chooses a selected peer in a `BTreeSet<Baseboard>`. For success, starting a node may not complete the connection before 
the [learner selects a peer](https://github.com/oxidecomputer/omicron/blob/9191af67b9aff710dac93ada6012b1e9b7d0c79c/bootstore/src/schemes/v0/fsm.rs#L287) or the Baseboard of the `non_learner_peer` may sort after the other peer. 

I think we should swap lines 1437 and 1438 so that the proper node is always chosen, but @luqmana claimed this causes hangs. I am going to dig into this in between BGP testing on madrid.

Reply:
I confirmed the hang and it makes sense. In order for a share to be handed out from one node to a learner, the encrypted 'extra shares' must be decrypted by recomputing the rack secret. If we don't start the second peer back up there's only one peer running and the rack secret cannot be recomputed, and a share is never handed out. The learner tries indefinitely and only returns once it learns a share. This is confirmed by the logs.

If we have connections at `learner-2` from peer `A` and peer `B`, the learner will always select A to learn from since it is sorted first the in the `peer.connections` BTreeSet. What this test was trying to do was to figure out the sort order dynamically and let the connections land where they may, then force a send to the same connection. We can simplify the test by always forcing the learner to pick peer A by asking the learner which connections it currently has and waiting until it has A, and only then trying to call `init_learner`. I'll make that change, PR it to this and then we should be good to go!